File: smalivm/src/main/java/org/cf/smalivm/emulate/MethodEmulator.java
package org.cf.smalivm.emulate;

import org.cf.smalivm.SideEffect;
import org.cf.smalivm.VirtualMachine;
import org.cf.smalivm.context.ExecutionContext;
import org.cf.smalivm.opcode.Op;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class MethodEmulator {

    private static final Logger log = LoggerFactory.getLogger(MethodEmulator.class.getSimpleName());
    private static Map<String, Class<? extends EmulatedMethod>> emulatedMethods = new HashMap<>();

    static {
        addMethod("Lorg/cf/simplify/Utils;->breakpoint()V", org_cf_simplify_Utils_breakpoint.class);
        addMethod("Ljava/lang/Class;->forName(Ljava/lang/String;)Ljava/lang/Class;", java_lang_Class_forName.class);
        addMethod("Ljava/lang/reflect/Field;->get(Ljava/lang/Object;)Ljava/lang/Object;", java_lang_reflect_Field_get.class);
        addMethod("Ljava/lang/Object;->getClass()Ljava/lang/Class;", java_lang_Object_getClass.class);
    }

    private final VirtualMachine vm;
    private final ExecutionContext context;
    private final String methodDescriptor;
    private final EmulatedMethod method;

    public MethodEmulator(VirtualMachine vm, ExecutionContext context, String methodDescriptor) {
        this.vm = vm;
        this.context = context;
        this.methodDescriptor = methodDescriptor;
        method = getMethod(methodDescriptor);
    }

    public static void addMethod(String methodDescriptor, Class<? extends EmulatedMethod> methodClass) {
        emulatedMethods.put(methodDescriptor, methodClass);
    }

    public static boolean canEmulate(String methodDescriptor) {
        return emulatedMethods.containsKey(methodDescriptor);
    }

    public static boolean canHandleUnknownValues(String methodDescriptor) {
        Class<? extends EmulatedMethod> methodClass = emulatedMethods.get(methodDescriptor);

        return (methodClass != null) && (UnknownValuesMethod.class.isAssignableFrom(methodClass));
    }

    public static void clearMethods() {
        emulatedMethods.clear();
    }

    private static EmulatedMethod getMethod(String methodDescriptor) {
        Class<? extends EmulatedMethod> methodClass = emulatedMethods.get(methodDescriptor);
        EmulatedMethod em = null;
        try {
            em = methodClass.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
            e.printStackTrace();
        }

        return em;
    }

    public void emulate(Op op) {
        try {
            if (method instanceof MethodStateMethod) {
                ((MethodStateMethod) method).execute(vm, op, context.getMethodState());
            } else {
                ((ExecutionContextMethod) method).execute(vm, op, context);
            }
        } catch (Exception e) {
            if (log.isWarnEnabled()) {
                log.warn("Unexpected JVM exception raised while emulating {}: ", methodDescriptor, e);
            }
        }
    }

    public SideEffect.Level getSideEffectLevel() {
        return method.getSideEffectLevel();
    }

    public Set<Throwable> getExceptions() {
        return method.getExceptions();
    }

}

File: smalivm/src/main/java/org/cf/smalivm/emulate/MethodEmulator.kt
package org.cf.smalivm.emulate

import org.cf.smalivm.type.VirtualMethod
import org.cf.smalivm.ExecutionNode
import org.cf.smalivm.ExecutionState
import org.cf.smalivm.UnresolvedChild
import org.cf.smalivm.VirtualMachine
import org.slf4j.LoggerFactory


class MethodEmulator {
    companion object {
        private val log = LoggerFactory.getLogger(MethodEmulator::class.java.simpleName)
        private val EMULATED_METHOD_SIGNATURES: MutableMap<String, EmulatedMethodCall> = HashMap()
        private val EMULATED_METHOD_MATCHES: MutableMap<(methodSignature: String) -> Boolean, EmulatedMethodCall> = HashMap()
        private val UNKNOWN_HANDLER_SIGNATURES: MutableSet<String> = HashSet()
        private val UNKNOWN_HANDLER_MATCHES: MutableSet<(methodSignature: String) -> Boolean> = HashSet()

        init {
            addMethod("Landroid/text/TextUtils;->isEmpty(Ljava/lang/CharSequence;)Z", android_text_TextUtils_isEmpty())
            addMethod("Lorg/cf/simplify/Utils;->breakpoint()V", org_cf_simplify_Utils_breakpoint())
            addMethod("Ljava/lang/Class;->forName(Ljava/lang/String;)Ljava/lang/Class;", java_lang_Class_forName())
            addMethod("Ljava/lang/reflect/Field;->get(Ljava/lang/Object;)Ljava/lang/Object;", java_lang_reflect_Field_get())
            addMethod("Ljava/lang/Object;->getClass()Ljava/lang/Class;", java_lang_Object_getClass())
            addMethod("Ljava/lang/Enum;-><init>(Ljava/lang/String;I)V", java_lang_Enum_init())
            addMethod("Ljava/lang/Object;-><init>()V", java_lang_Object_init())
            addMethod({ signature: String -> signature.endsWith("->clone()Ljava/lang/Object;") }, java_lang_Object_clone())
        }

        private fun addMethod(match: (methodSignature: String) -> Boolean, method: EmulatedMethodCall) {
            EMULATED_METHOD_MATCHES[match] = method
            if (method.canHandleUnknownValues) {
                UNKNOWN_HANDLER_MATCHES.add(match)
            }
        }

        fun addMethod(methodSignature: String, method: EmulatedMethodCall) {
            EMULATED_METHOD_SIGNATURES[methodSignature] = method
            if (method.canHandleUnknownValues) {
                UNKNOWN_HANDLER_SIGNATURES.add(methodSignature)
            }
        }

        fun canEmulate(method: VirtualMethod): Boolean {
            return canEmulate(method.signature)
        }

        fun canEmulate(methodSignature: String): Boolean {
            if (EMULATED_METHOD_SIGNATURES.containsKey(methodSignature)) {
                return true
            }
            for ((match, method) in EMULATED_METHOD_MATCHES) {
                if (match.invoke(methodSignature)) {
                    // Cache this signature for faster lookups later
                    addMethod(methodSignature, method)
                    return true
                }
            }
            return false
        }

        fun canHandleUnknownValues(method: VirtualMethod): Boolean {
            return canHandleUnknownValues(method.signature)
        }

        fun canHandleUnknownValues(methodSignature: String): Boolean {
            return UNKNOWN_HANDLER_SIGNATURES.contains(methodSignature)
        }

        fun clearMethods() {
            EMULATED_METHOD_SIGNATURES.clear()
        }

//        private fun getMethod(methodSignature: String): EmulatedMethodCall {
//            val methodClass = EMULATED_METHOD_SIGNATURES[methodSignature]!!
//            return try {
//                methodClass.newInstance()
//            } catch (e: Exception) {
//                throw RuntimeException("Exception instantiating emulated method $methodSignature", e)
//            }
//        }

        fun emulate(method: VirtualMethod, state: ExecutionState, callerNode: ExecutionNode?, vm: VirtualMachine): Array<out UnresolvedChild> {
            return emulate(method.signature, state, callerNode, vm)
        }

        fun emulate(methodSignature: String, state: ExecutionState, callerNode: ExecutionNode?, vm: VirtualMachine): Array<out UnresolvedChild> {
            val emulatedMethodCall = EMULATED_METHOD_SIGNATURES[methodSignature]!!
            return try {
                emulatedMethodCall.execute(state, callerNode, vm)
            } catch (e: Exception) {
                throw RuntimeException("Unhandled exception thrown emulating $methodSignature", e)
            }
        }
    }
}
